WordPress/Classes/Utility/AuthTokenIssueSolver.swift

```diff
@@ -0,0 +1,74 @@
+import UIKit
+import WordPressData
+
+struct AuthTokenIssueSolver {
+    private let coreData: CoreDataStack
+
+    init(coreData: CoreDataStack = ContextManager.shared) {
+        self.coreData = coreData
+    }
+
+    /// - note: The completion callback is run immediatelly if there is no issue.
+    func fixAuthTokenIssueIfNeeded(in window: UIWindow, _ completion: @escaping () -> Void) {
+        guard hasAuthTokenIssues() else {
+            completion()
+            return
+        }
+
+        let signInVC = WordPressAuthenticationManager.signinForWPComFixingAuthToken { cancelled in
+            if cancelled {
+                // We present asynchronously to prevent an issue where the Login VC would dismiss the
+                // alert instead of itself.
+                DispatchQueue.main.async {
+                    showCancelReAuthenticationAlert(in: window, onDeletionConfirmed: {
+                        let accountService = AccountService(coreDataStack: ContextManager.sharedInstance())
+                        accountService.removeDefaultWordPressComAccount()
+                        completion()
+                    })
+                }
+            } else {
+                completion()
+            }
+        }
+
+        window.rootViewController = signInVC
+
+        showExplanationAlertForReAuthentication(in: signInVC)
+    }
+
+    /// Call this method to know if the local installation of WPiOS has the
+    /// authToken issue this class was designed to solve.
+    /// - returns: `true` if the local WPiOS installation needs to be fixed by this class.
+    private func hasAuthTokenIssues() -> Bool {
+        guard let account = try? WPAccount.lookupDefaultWordPressComAccount(in: coreData.mainContext) else {
+            return false
+        }
+        return account.authToken == nil
+    }
+
+    private func showCancelReAuthenticationAlert(in window: UIWindow, onDeletionConfirmed: @escaping () -> Void) {
+        let title = NSLocalizedString("Careful!", comment: "Title for the warning shown to the user when he refuses to re-login when the authToken is missing.")
+        let message = NSLocalizedString("Proceeding will remove all WordPress.com data from this device, and delete any locally saved drafts. You will not lose anything already saved to your WordPress.com blog(s).", comment: "Message for the warning shown to the user when he refuses to re-login when the authToken is missing.")
```

Comment:
Would be good to add keys while we're at it.

```suggestion
        let title = NSLocalizedString(
            "auth_token_issue_solver.cancel_alert.title",
            value: "Careful!",
            comment: "Title for the warning shown to the user when he refuses to re-login when the authToken is missing."
        )
        let message = NSLocalizedString(
            "auth_token_issue_solver.cancel_alert.message",
            value: "Proceeding will remove all WordPress.com data from this device, and delete any locally saved drafts. You will not lose anything already saved to your WordPress.com blog(s).",
            comment: "Message for the warning shown to the user when he refuses to re-login when the authToken is missing."
        )
```